`CodeBlock` docs (HDS-2595) (#1749)
hashicorp/design-system: website/docs/components/code-block/partials/guidelines/guidelines.md

## Usage

### When to use

- When displaying code examples and longer snippets of code that benefit from syntax highlighting.

### When not to use

- As a full-blown code editor

> - When displaying keys, tokens, or other generated values, use a [CopySnippet](/components/copy/snippet) instead.

Reviewer: https://github.com/hashicorp/design-system/pull/1691/files#r1367333392

Reviewer 2: Resolved in https://github.com/hashicorp/design-system/pull/1749/commits/19a092530be5f2419739f829a8d000a99a7cb4a7

Reviewer 2: I also added a section for when to use the CopySnippet vs CodeBlock, trying to think of more examples, what do you think?

Reviewer: I'll add a comment there. In the meantime I'm resolving this one.

## Inline

> The `isInline` property increases the portability of the CodeBlock to ensure that it can be used in different contexts. For example, a common use case of the CodeBlock is in a "standalone" context, which can be part of a form, multi-step process, and is generally a part of the normal layout flow.

Reviewer: https://github.com/hashicorp/design-system/pull/1691/files#r1367334741

Reviewer 2: Resolved in https://github.com/hashicorp/design-system/pull/1749/commits/19a092530be5f2419739f829a8d000a99a7cb4a7

Sometimes it may be necessary to use the CodeBlock in a more dense layout or nested within another component. In this circumstance, setting `isInline` to true ensures that the CodeBlock fits alongside other elements, in a split view, or as part of a larger layout mechanism.

![Code Block in a block context](/assets/components/code-block/code-block-block-level.png)

## Header

Use a `title` or `description` in the Header to provide additional information, instructions, or to label a CodeBlock. Both of these properties are optional, but including them can help to provide additional context about a specific block of code.

![Example metadata in the Code Block](/assets/components/code-block/code-block-metadata.png)

### When not to use a Header

There can be an overlap between content that you may choose to include in the Header as a `title` or `description`, and content that is part of the normal layout flow in a headline or paragraph. If it is necessary to elevate this content in the hiearchy of the page, we recommend including it in the normal layout flow, rather than as a `title` or `description` within the Code Block.

> ![An example showcasing the Code Block paired with content in the nautral flow](/assets/components/code-block/code-block-dont-use-metadata.png)

Reviewer: see https://github.com/hashicorp/design-system/pull/1691/files#r1367338989

I suspect some changes that @jorytindall did went lost

Reviewer 2: Resolved in https://github.com/hashicorp/design-system/pull/1749/commits/19a092530be5f2419739f829a8d000a99a7cb4a7

Author: Looks like that. Thanks for updating this branch, Jory! 🙏

## CopyButton

Use a CopyButton within the CodeBlock to make copying the snippet a single action. More details can be found in the [CopyButton guidelines](/components/copy/button).

> ![An example of the Copy Button within the Code Block](/assets/components/code-block/code-block-copy-button.png)

Reviewer: https://github.com/hashicorp/design-system/pull/1691/files#r1367339696

Reviewer 2: Resolved in https://github.com/hashicorp/design-system/pull/1749/commits/19a092530be5f2419739f829a8d000a99a7cb4a7

### Line selection

> If a user needs to copy only a portion of the CodeBlock, the relevant portion can be selected with a cursor and copied via the keyboard or mouse.

Reviewer: @alex-ju @jorytindall we need to take a decision on this (foreground/background selection colors)
https://github.com/hashicorp/design-system/pull/1691/files#r1367341752

Reviewer 2: I responded to this and can update the image based on the decision.

Reviewer 2: Decision: use a `green` background and a foreground/text color that is inverse of the backgroud [as seen here](https://www.figma.com/file/1TLwCzn7Bjkhw6OmCTqbPN/%F0%9F%92%80-Graveyard---FY24-Q3%2FQ4?type=design&node-id=52%3A336750&mode=design&t=8nJXUPC6H0N0rOrV-1). I'll update the image.

Author: The [equivalent change in code ](https://github.com/hashicorp/design-system/pull/1687/commits/764909cc2929651bcb9f2f48d9f25f69b82fa2e9).

### Applying syntax highlighting

If you wish to create custom examples using the Code Block, we publish all of the relevant syntax highlighting styles in the HDS [Foundations](https://www.figma.com/file/oQsMzMMnynfPWpMEt91OpH/HDS-Product---Foundations?type=design&node-id=2130%3A2&mode=design&t=Pfj7CheLS6cR0hKa-1) library. However, due to the number languages supported by the component, the color styles use a generic naming schema (e.g., cyan, red, purple) to remain as agnostic as possible when being applied to different languages.

!!! Info

Working directly with an engineering partner can reveal exactly how a snippet will render in the component and should be the first course of action when creating custom snippets. Understanding Prism's [token hierarchy](https://prismjs.com/tokens.html) can also be helpful when creating examples.

If you have questions or need assistance creating custom examples, don’t hesitate to reach out the the HDS team for [support](/about/support).
!!!

> To aid in understanding how the highlighting theme is applied via Prism's tokens, we've provided a high-level, non-exhaustive list of token names and how they might be applied depending on the syntax.

Reviewer: https://github.com/hashicorp/design-system/pull/1691/files#r1367345926

Reviewer 2: Resolved in https://github.com/hashicorp/design-system/pull/1749/commits/19a092530be5f2419739f829a8d000a99a7cb4a7

## Usage

### When to use

- When displaying code examples and longer snippets of code that benefit from syntax highlighting.

### When not to use

- As a full-blown code editor.

### When to use a CopySnippet vs a CodeBlock

> There is some overlap in the copying functionality of the CopySnippet and the CodeBlock. Which to use generally comes down to the complexity of the code/value displayed within the component, and whether the user benefits from seeing the larger context of the code example.

Reviewer: ```suggestion
There is some overlap in the copying functionality of the `CopySnippet` and the `CodeBlock`. Which to use generally comes down to the complexity of the code/value displayed within the component, and whether the user benefits from seeing the larger context of the code example.
```

> For example; if allowing the user to copy an API key or other **singular** value where they don’t necessarily need to see the entire context of where the copied value is expected to be pasted into, use a CopySnippet.

Reviewer: I _think_ this paragraph needs some refinement, is convoluted and hard to follow. I would use the same format used below: "Use a `CopySnippet` if:"

> Use a CodeBlock if:

Reviewer: ```suggestion
Use a `CodeBlock` if:
```

> - The user benefits from seeing the larger context of where a value should be pasted into, e.g., in a configuration file.

Reviewer: I am not sure the code-block allows to "seeing the larger context of where a value should be pasted into, e.g., in a configuration file".

For me the main difference is this one:  the main action for a code block is seeing the code, the copy is secondary; for a copy snippet is the other way round.

Reviewer 2: That's a great breakdown, thanks for that!

> - If the example being copied is more than a single line.

Reviewer: This is also true for copy snippets (I can share examples, if you want)

Reviewer 2: Ok, I've removed this for now

## Usage

### When to use

- When displaying code examples and longer snippets of code that benefit from syntax highlighting.

### When not to use

- As a full-blown code editor.

> ### When to use a CodeBlock vs a CopySnippet

Reviewer: On the [other documentation page](https://helios.hashicorp.design/components/copy/snippet), is called "Copy Snippet"
```suggestion
### When to use a CodeBlock vs a Copy Snippet
```

> ### When to use a CodeBlock vs a Copy Snippet

Reviewer 3: [suggestion] This section may benefit from some visual do/don't examples.

Reviewer 2: Ok, I've added a couple of examples. I think they're probably representative enough, but I'm not 100% sold on them. WDYT?

Reviewer: Personally, I don't think adding do/don't examples is adding much clarity. There is an intrinsic ambiguity between the usage of two components (_"There is some overlap in the copying functionality of the Copy Snippet and the CodeBlock"_) and for me the visual examples add to the confusion, instead of providing clarity. We should trust the consumers to make the right choice, in their context, since we can't ourselves pinpoint to specific criteria on how to choose which.

Reviewer 2: Agreed, I'll remove them.

> There is some overlap in the copying functionality of the CopySnippet and the CodeBlock. Which to use generally comes down to the complexity of the code/value displayed within the component, and whether the user benefits from seeing the larger context of the code example.

Reviewer: ```suggestion
There is some overlap in the copying functionality of the [Copy Snippet](/components/copy/snippet) and the CodeBlock. Which to use generally comes down to the complexity of the code/value displayed within the component, and whether the user benefits from seeing the larger context of the code example.
```

> Use a CodeBlock if:

Reviewer: Should the "if" be removed here, since the following paragraphs also start with an "if" (or viceversa) 
```suggestion
Use a CodeBlock:
```

Same for the block below.

Reviewer 2: Yep, that's a good callout, thank you!

> Use a CodeBlock:

Reviewer 3: [suggestion] I wonder if if might be easier to scan if it was bold?

- If viewing or displaying code is the primary purpose, and copying the code is secondary.
- If the example consists of a command, e.g., a `curl` and `bash` script. These are oftentimes on a single line, but consist of multiple commands and functions.

> Use a CopySnippet if:

Reviewer: ```suggestion
Use a Copy Snippet if:
```

> Use a Copy Snippet:

Reviewer 3: [suggestion] I wonder if if might be easier to scan if it was bold?

- If copying the code is the primary purpose, and viewing the code is secondary.
- If allowing the user to copy an API key or other **single** value or string.
- If seeing the value in a larger context of where it's expected to be pasted into isn’t necessary.

> ## Standalone

Reviewer 3: Great idea to put this towards the top!

The `isStandalone` property increases the portability of the CodeBlock to ensure that it can be used in different contexts. For example, a common use case of the CodeBlock is in a "standalone" context, which can be part of a form, multi-step process, and is generally a part of the normal layout flow.

![Code Block with rounded corners in a standalone context](/assets/components/code-block/code-block-rounded-standalone.png)

Sometimes it may be necessary to use the CodeBlock in a more dense layout or nested within another component. In this circumstance, setting `isStandalone` to false ensures that the CodeBlock fits alongside other elements, in a split view, or as part of a larger layout mechanism.

![Code Block in a block context](/assets/components/code-block/code-block-block-level.png)

## Header

> Use a `title` or `description` in the header to provide additional information, instructions, or to label a CodeBlock. Both of these properties are optional, but including them can help to provide additional context about a specific block of code.

Reviewer: ```suggestion
Use a `title` and/or `description` in the header to provide additional information, instructions, or to label a CodeBlock. Both of these properties are optional, but including them can help to provide additional context about a specific block of code.
```

![Example metadata in the Code Block](/assets/components/code-block/code-block-metadata.png)

### When not to use a header

> There can be an overlap between content that you may choose to include in the header as a `title` or `description`, and content that is part of the normal layout flow in a headline or paragraph. If it is necessary to elevate this content in the hierarchy of the page, we recommend including it in the normal layout flow, rather than as a `title` or `description` within the Code Block.

Reviewer: ```suggestion
There can be an overlap between content that you may choose to include in the header as a `title` or `description`, and content that is part of the normal layout flow in a headline or paragraph. If it is necessary to elevate this content in the hierarchy of the page, we recommend including it in the normal layout flow, rather than as a `title` or `description` within the CodeBlock.
```

The **Ember** component uses [Prism.js](https://prismjs.com/index.html) to handle syntax highlighting and comes with a pre-defined set of languages.

In **Figma** we provide a handful of example languages that are intended to be _representative_ of the end result in production. Syntax highlighting in Figma is a non-trivial process and requires the manual application of color styles to each "type" of code. Despite this, creating a custom code snippet with the Code Block is supported by typing/pasting into the text layer, but syntax highlighting will not be automatically applied.

### Applying syntax highlighting

If you wish to create custom examples using the Code Block, we publish all of the relevant syntax highlighting styles in the HDS [Foundations](https://www.figma.com/file/oQsMzMMnynfPWpMEt91OpH/HDS-Product---Foundations?type=design&node-id=2130%3A2&mode=design&t=Pfj7CheLS6cR0hKa-1) library. However, due to the number of languages supported by the component, the color styles use a generic naming schema (e.g., cyan, red, purple) to remain as agnostic as possible when being applied to different languages.

> For more details around syntax visit the [specifications](?tab=specifcations).

Reviewer: ```suggestion
For more details around syntax visit the [specifications](?tab=specifications).
```
